src/14-curated-discovery/helpers.ts

```diff
@@ -75,3 +78,82 @@ export async function getImageDataForArtworks(
   const imageData = _.fromPairs(imageFieldPairs)
   return imageData
 }
+
+/**
+ * Use multimodal models’ computer vision capabilities to
+ * describe an artwork image.
+ *
+ * @param args.url URL of the artwork image that is to be described
+ * @param args.model Model to use, as defined by Vercel AI SDK
+ * @param args.maxWords Maximum number of words for prose description of the image (default: 200)
+ * @returns An object that describes the image
+ */
+export async function describeImage(args: {
+  url: string
+  model: Parameters<typeof generateObject>[0]["model"]
+  maxWords?: number
+}) {
+  const { url, model } = args
+  const maxWords = args.maxWords || 200
+
+  const VALID_COLORS = [
+    "red",
+    "orange",
+    "yellow",
+    "green",
+    "blue",
+    "purple",
+    "brown",
+    "gray",
+    "pink",
+  ]
+
+  const system = dedent`
+    You are an expert on art and your task is to provide detailed
+    descriptions of artworks.
+
+    Consider the visual qualities of the artwork, as well as
+    the artwork's relation to art historical styles and movements.
+
+    If you detect any objects in the image, include them in your output.
+
+    If you detect any words in the image, include them in your output.
+
+    Include up to 3 prominent colors in your output, excluding any background paper, canvas, or wall colors. The only valid colors are ${VALID_COLORS.join(", ")}.
+  `
+
+  const schema = z.object({
+    description: z.string(),
+    objects: z.string().array().optional(),
+    words: z.string().array().optional(),
+    colors: z.string().array().max(3),
+  })
```

Comment:
Nice. Wasn't aware of these options. 

I see the [openAi docs say](https://platform.openai.com/docs/guides/text-generation/json-mode) `JSON mode will not guarantee the output matches any specific schema, only that it is valid and parses without errors.`. [Anthropic](https://docs.anthropic.com/en/docs/control-output-format) is less clear, but wondering if you noticed it seems to conform to the provided schema in most cases?

Reply:
Mostly just works. Very rarely I saw output that instead of valid json, was instead Markdown with code-fenced json, i.e.

Expected:
```json-c
{ "foo" : "bar" }
```

Actual:
````md
```json
{ "foo" : "bar" }
```
````

I also had some trouble when specifying `VALID_COLORS` as a Zod enum in the schema — the models still sometimes returned a color that wasn't in the valid list, so I relaxed the schema to allow any color names, but _prompted_ it to use only valid ones, and hoped for the best 🤞🏽